### PR TITLE
GCC compilers version 8 or newer are highly recommended.

### PR DESCRIPTION
**Pull request by slaff**
Start warning when older compiler is used.
Middle of 2021 we should enforce the usage of newer compilers only.

### CODE REVIEW

File: Sming/build.mk [hunk 1]
```diff
@@ -189,7 +189,15 @@ SMING_CXX_STD		?= c++17
...
-
+GCC_MIN_MAJOR_VERSION = 8
+GCC_VERSION_COMPATIBLE := $(shell expr `$(CC) -dumpversion | cut -f1 -d.` \>= $(GCC_MIN_MAJOR_VERSION))
+
+ifeq ($(GCC_VERSION_COMPATIBLE),0)
+$(warning ***** Please, upgrade your GCC compiler to version $(GCC_MIN_MAJOR_VERSION) or newer *************)
+ifneq ($(GCC_UPGRADE_URL),)
+$(info Instructions for upgrading your compiler can be found here: $(GCC_UPGRADE_URL))
+endif 
+endif
```
Comment (mikee47) [inline Sming/build.mk:200]: How about:

```
GCC_MIN_VERSION := "8.0"
GCC_VERSION_COMPATIBLE := $(shell expr "$(GCC_VERSION)" ">=" "$(GCC_MIN_VERSION)")

ifeq ($(GCC_VERSION_COMPATIBLE),0)
$(warning ***** Please, upgrade your GCC compiler to version $(GCC_MIN_VERSION) or newer *************)
ifneq ($(GCC_UPGRADE_URL),)
$(info Instructions for upgrading your compiler can be found here: $(GCC_UPGRADE_URL))
endif 
endif
```

https://stackoverflow.com/questions/42419301/makefile-emit-warning-gcc-version-is-lower-than-4-8-0

File: README.md [hunk 2]
```diff
@@ -74,6 +74,15 @@ Please note Version 4 documentation is at [sming.readthedocs.io](https://sming.r
...
 
+
+### GCC compiler >= 8
+
+Your GCC compiler should be version 8 or newer! See the different architectures for details.
+Newer compilers
+
+- produce smaller and better code
+- have features which improve code quality, easy of use, etc.
+
```
Comment (mikee47) [inline README.md:85]: Bug fixes!

File: Sming/Arch/Esp8266/build.mk [hunk 3]
```diff
@@ -40,6 +40,8 @@ OBJCOPY			 	:= $(TOOLSPEC)objcopy
 OBJDUMP			 	:= $(TOOLSPEC)objdump
 GDB				:= $(TOOLSPEC)gdb
 
+GCC_UPGRADE_URL := https://sming.readthedocs.io/en/latest/arch/esp8266/getting-started/eqt.html
```
Comment (mikee47) [inline Sming/Arch/Esp8266/build.mk:43]: Should we also update the 'pre-release' note on this page? I've yet to find any issues with it :-)

**Comment by slaff** [comment]
@mikee47 Recommended changes were applied. Something else to add to this PR?

**Review by mikee47 — APPROVED** [review]
Great, thanks slaff. I know there have been further updates to the esp-quick-toolchain since we introduced it but seems pretty solid to me! At some point we should take a look at what's changed.

**Comment by slaff** [comment]
> At some point we should take a look at what's changed.

Absolutely! Up to now there are only pre-release versions between `3.0.0-gnu5` and the latest `3.0.0-newlib330_gnu14`.

**Comment by frankdownunder** [comment]
i have a version 10 compiler
it fails the comparison

this makefile;
GCC_VERSION			:= $(shell $(CC) -dumpversion)
GCC_MIN_VERSION := "8.0"
GCC_VERSION_COMPATIBLE := $(shell expr "$(GCC_VERSION)" ">=" "$(GCC_MIN_VERSION)")
all:
	@echo "GCC_VERSION="$(GCC_VERSION)
	@echo GCC_MIN_VERSION=$(GCC_MIN_VERSION)
	@echo GCC_VERSION_COMPATIBLE=$(GCC_VERSION_COMPATIBLE)


prints:

GCC_VERSION=10
GCC_MIN_VERSION=8.0
GCC_VERSION_COMPATIBLE=0

**Comment by slaff** [comment]
@frankdownunder  Go to the application that you are trying to compile and type
```
make list-config
```
Then paste here the whole output.

**Comment by mikee47** [comment]
It's because the comparison is lexical, "1" < "8". Need to strip out everything but the major part then compare numerically i.e. `$(shell $$(($(GCC_VERSION_MAJOR) >= $(GCC_MIN_VERSION_MAJOR))))`

**Comment by slaff** [comment]
@frankdownunder Can you test if this PR: https://github.com/SmingHub/Sming/pull/2172 is fixing the problem?

**Pull request merged**